---
title: "RNAseq Post Analysis Visualization"
author: "Evan Udine"
date: "`r Sys.Date()`"
output: 
  rmarkdown::html_document: 
   code_folding: show
   toc: true
   toc_float: true
   smooth_scroll: true
   number_sections: false
   self_contained: true
---
#Set Up
##Libraries
```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
sessionInfo()
suppressPackageStartupMessages(library(data.table))
suppressPackageStartupMessages(library(ggplot2))
suppressPackageStartupMessages(library(ggrepel))
suppressPackageStartupMessages(library(ggfortify))
suppressPackageStartupMessages(library(gplots))
suppressPackageStartupMessages(library(RColorBrewer))
```

#Distributions
```{r Distributions}
res = read.table("/sc/hydra/projects/pd-omics/evan/MyND-Analysis/data/results/rsem.mynd.dte.txt", header = T)
head(res)
p = ggplot(res, aes(P.Value))
p + geom_density(color="darkblue", fill="lightblue") + theme_bw() + ggtitle("P Value Distribution")

p = ggplot(res, aes(adj.P.Val))
p + geom_density(color="darkblue", fill="lightblue") + theme_bw() + ggtitle("FDR Distribution")

p = ggplot(res, aes(logFC))
p + geom_density(color="darkblue", fill="lightblue") + theme_bw() + ggtitle("Fold Change Distribution")
```

#DE Plots
```{r DE Plots}
plot.data = res
head(plot.data)
plot.data$id = rownames(plot.data)
data = data.frame(plot.data)
data$P.Value = -log10(data$P.Value)
head(data)
table(res$adj.P.Val < .05)
data$fifteen = as.factor(abs(data$adj.P.Val < 0.05))

data = data[order(-data$adj.P.Val),]
#MA Plot
ma = ggplot(data, aes(AveExpr, logFC, colour = fifteen))
ma + geom_point() +
  scale_color_manual(values = c("black", "red"), labels = c(" > 0.15", "< 0.15")) +
  labs(title = "MA Plot", color = "FDR") +
  theme(plot.title = element_text(hjust = 0.5)) + ylim(-3, 3) + xlim(-4, 14)

#Volcano Plot
vp = ggplot(data, aes(logFC, P.Value, colour = fifteen))
vp + geom_point() +
  scale_color_manual(values = c("black", "red"), labels = c("> 0.05", "< 0.05")) + 
  labs(title = "Gene Level Volcano Plot", color = "FDR") +
  theme(plot.title = element_text(hjust = 0.5)) +
  xlim(-2.5,2.5) + ylim(0,10) + ylab("-log10 pvalue")

ann = read.table('/sc/hydra/projects/pd-omics/GRCH38.p12/ens.geneid.gencode.v30', row.names = 1,  header = T, check.names = F)
head(ann)

merged_ann = merge(res, ann, by  = 0, sort = F)
head(merged_ann)
```

#Heatmaps {.tabset .tabset-fade .tabset-pills}
```{r heatmaps, results='asis'} 
#Inlclude heatmap of top 100 Differentially Expressed Genes
x = read.table('/sc/hydra/projects/pd-omics/mynd/analyses/analysis_9_19_19/differential_expression_11.04.2019/rsem.mynd.design_7.txt', header = T, check.names  = F)
dim(x)

#Select ids top 50 genes
topdge = x[c(1:100),]
#
topdge = topdge[c(1)]

#read in expression matrix
y = read.table('/sc/hydra/projects/pd-omics/mynd/analyses/analysis_9_19_19/mynd.isoforms', header = T, check.names = F)
#head(y)
tot =  merge(res, ann, by  = 0, sort = F)
dim(y)

pheno = read.table('/sc/hydra/projects/pd-omics/mynd/analyses/analysis_9_19_19/mynd_metadata_9_19_19.txt', header = T, row.names = 1)
pheno = pheno[order(pheno$status),]
pheno$rna_batch = as.factor(pheno$rna_batch)
total = merge(topdge, y, by = 0,  sort = F)
dim(total)

total$Row.names = NULL
total$logFC = NULL

#scale data
logTransformed.dat = log2(total + 1)
#head(logTransformed.dat)

z.mat <- t(scale(t(logTransformed.dat), scale=TRUE, center=TRUE))

tz.mat = data.frame(t(z.mat), check.names = F)
tz.mat = tz.mat[rownames(pheno),]
z.mat = as.matrix(t(tz.mat))
identical(colnames(z.mat), rownames(pheno))

col.status <- c('grey','black')[pheno$status]
my_palette <- rev(brewer.pal(9, "RdBu"))

#Run heatmap2 for visualization - No covariate Regression
heatmap.2(
  z.mat,
  col = my_palette,
  Rowv = T,
  Colv = F,
  trace = "none",
  scale = "none",
  adjRow = c(0, 0.5),
  margins = c(6, 10),
  key = TRUE,
  keysize = 1.3,
  xlab = "Samples",
  ylab = "Genes",
  ColSideColors = col.status)
title(main = "Gene Level Top 100 DGE")

#regress out covariates
getResiduals <- function(exprLine) {
  ourCovRegression <- lm(exprLine ~ pheno$rna_batch + pheno$PCT_USABLE_BASES + pheno$PCT_RIBOSOMAL_BASES + pheno$rin_value + pheno$age + pheno$mds1 + pheno$mds2 + pheno$mds3 + pheno$mds4 + pheno$gender);
  ourResiduals <- ourCovRegression$residuals;
  return(ourResiduals)
}
allResiduals <- data.frame(t(apply(total, 1, getResiduals)), check.names = F)
resid = as.matrix(allResiduals)
#head(resid)
#scale data
logTransformed.dat = log2(total + 1)
#head(logTransformed.dat)
z.matr <- t(scale(t(resid), scale=TRUE, center=TRUE))

tz.matr = data.frame(t(z.matr), check.names = F)
tz.matr = tz.matr[rownames(pheno),]
z.matr = as.matrix(t(tz.matr))
identical(colnames(z.matr), rownames(pheno))

col.status <- c('grey','black')[pheno$status]
my_palette <- rev(brewer.pal(9, "RdBu"))

#Run heatmap2 for visualization - Regressing out covariates from tpms
heatmap.2(
  z.matr,
  col = my_palette,
  Rowv = T,
  Colv = F,
  trace = "none",
  scale = "none",
  adjRow = c(0, 0.5),
  margins = c(6, 10),
  key = TRUE,
  keysize = 1.3,
  xlab = "Samples",
  ylab = "Genes", 
  ColSideColors = col.status)
title(main = "Gene Level Top 100 DGE")
```
